### PR TITLE
[Keyboard] Add Ky01

### PR DESCRIPTION
**Pull request by mcoirault**
Add support for the KY01 with default qmk and VIA.

[GH link](https://geekhack.org/index.php?topic=107809.0)

## Description

Adding support for the ky01. The PCB designer never merged their work to the main qmk repo, so this PR attempts to fix that.

My commits add via support and modify the original designer's work to match the guidelines to the best of my ability.

The firmware was built from my branch and distributed on the discord server of the ky01 project. No issue was reported.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Review by ridingqwerty — APPROVED** [review]

File: keyboards/ky01/config.h [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+//#define MAGIC_KEY_LAYER8         8
+//#define MAGIC_KEY_LAYER9         9
+//#define MAGIC_KEY_BOOTLOADER     B
+//#define MAGIC_KEY_BOOTLOADER_ALT ESC
+//#define MAGIC_KEY_LOCK           CAPS
+//#define MAGIC_KEY_EEPROM         E
+//#define MAGIC_KEY_EEPROM_CLEAR   BSPACE
+//#define MAGIC_KEY_NKRO           N
+//#define MAGIC_KEY_SLEEP_LED      Z
+
```
Comment (fauxpark) [inline keyboards/ky01/config.h:176]:
```suggestion
```

File: keyboards/ky01/config.h [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+#define LCD_DATA2_PIN    6            //< pin for 4bit data bit 2
+#define LCD_DATA3_PIN    7            //< pin for 4bit data bit 3
+#define LCD_RS_PORT      LCD_PORT     //< port for RS line
+#define LCD_RS_PIN       3            //< pin  for RS line
+#define LCD_RW_PORT      LCD_PORT     //< port for RW line
+#define LCD_RW_PIN       2            //< pin  for RW line
+#define LCD_E_PORT       LCD_PORT     //< port for Enable line
+#define LCD_E_PIN        1            //< pin  for Enable line
+#endif
+*/
```
Comment (fauxpark) [inline keyboards/ky01/config.h:246]:
```suggestion
```

File: keyboards/ky01/keymaps/default/keymap.c [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include QMK_KEYBOARD_H
+
+#define ___ KC_TRNS
+
```
Comment (fauxpark) [inline keyboards/ky01/keymaps/default/keymap.c:19]:
```suggestion
```
This is already defined in QMK core, with seven underscores.

File: keyboards/ky01/keymaps/default/keymap.c [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        KC_LSFT,  KC_NUBS, KC_Z,    KC_X,  KC_C,  KC_V,  KC_B,     KC_N,  KC_M,  KC_COMMA, KC_DOT,  KC_SLSH,  KC_RSFT,  KC_UP,
+        KC_LCTL,  KC_LGUI, KC_LALT,                      KC_SPACE,               KC_RGUI,  KC_RALT, KC_RCTL,  KC_LEFT,  KC_DOWN,  KC_RIGHT
+    ),
+    [L1] = LAYOUT(
+        RESET,    ___, ___, ___, ___,      ___, ___, ___, ___, ___, ___, ___, ___, ___,
+        ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___,
+        ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___,      ___,
+        ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___,      ___,
+        ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___,
+        ___, ___, ___,                ___,           ___, ___, ___, ___, ___, ___
```
Comment (fauxpark) [inline keyboards/ky01/keymaps/default/keymap.c:41]:
```suggestion
        RESET,   _______, _______, _______, _______,          _______, _______, _______, _______, _______, _______, _______, _______, _______,
        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______,
        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______,
        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
        _______, _______, _______,                            _______,                   _______, _______, _______, _______, _______, _______
```

File: keyboards/ky01/keymaps/via/keymap.c [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include QMK_KEYBOARD_H
+
+#define ___ KC_TRNS
+
```
Comment (fauxpark) [inline keyboards/ky01/keymaps/via/keymap.c:19]:
```suggestion
```

File: keyboards/ky01/keymaps/via/keymap.c [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___,
+        ___, ___, ___,                ___,           ___, ___, ___, ___, ___, ___
+    ),
+    [L3] = LAYOUT(
+        ___,      ___, ___, ___, ___,      ___, ___, ___, ___, ___, ___, ___, ___, ___,
+        ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___,
+        ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___,      ___,
+        ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___,      ___,
+        ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___, ___,
+        ___, ___, ___,                ___,           ___, ___, ___, ___, ___, ___
```
Comment (fauxpark) [inline keyboards/ky01/keymaps/via/keymap.c:60]:
```suggestion
        _______,          _______, _______, _______, _______,          _______, _______, _______, _______, _______, _______, _______, _______, _______,
        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______,
        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______,
        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
        _______, _______, _______,                            _______,                   _______, _______, _______, _______, _______, _______
    ),
    [L2] = LAYOUT(
        _______,          _______, _______, _______, _______,          _______, _______, _______, _______, _______, _______, _______, _______, _______,
        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______,
        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______,
        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
        _______, _______, _______,                            _______,                   _______, _______, _______, _______, _______, _______
    ),
    [L3] = LAYOUT(
        _______,          _______, _______, _______, _______,          _______, _______, _______, _______, _______, _______, _______, _______, _______,
        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______,
        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______,
        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
        _______, _______, _______,                            _______,                   _______, _______, _______, _______, _______, _______
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/ky01/config.h (hunk 1 above)
Comment (mcoirault) [reply keyboards/ky01/config.h]: Suggestion applied in commit https://github.com/qmk/qmk_firmware/pull/12844/commits/45da15b3923e998097f6368e754b11232dc655dd

File: keyboards/ky01/config.h (hunk 2 above)
Comment (mcoirault) [reply keyboards/ky01/config.h]: Suggestion applied in commit https://github.com/qmk/qmk_firmware/pull/12844/commits/45da15b3923e998097f6368e754b11232dc655dd

File: keyboards/ky01/keymaps/default/keymap.c (hunk 3 above)
Comment (mcoirault) [reply keyboards/ky01/keymaps/default/keymap.c]: Suggestion applied in commit https://github.com/qmk/qmk_firmware/pull/12844/commits/45da15b3923e998097f6368e754b11232dc655dd

File: keyboards/ky01/keymaps/default/keymap.c (hunk 4 above)
Comment (mcoirault) [reply keyboards/ky01/keymaps/default/keymap.c]: Suggestion applied in commit https://github.com/qmk/qmk_firmware/pull/12844/commits/45da15b3923e998097f6368e754b11232dc655dd

File: keyboards/ky01/keymaps/via/keymap.c (hunk 5 above)
Comment (mcoirault) [reply keyboards/ky01/keymaps/via/keymap.c]: Suggestion applied in commit https://github.com/qmk/qmk_firmware/pull/12844/commits/45da15b3923e998097f6368e754b11232dc655dd

File: keyboards/ky01/keymaps/via/keymap.c (hunk 6 above)
Comment (mcoirault) [reply keyboards/ky01/keymaps/via/keymap.c]: Suggestion applied in commit https://github.com/qmk/qmk_firmware/pull/12844/commits/45da15b3923e998097f6368e754b11232dc655dd

**Comment by mcoirault** [comment]
Apologies for the clumsy solving of the requested changes, there might have been a way to batch them all up together from the github UI. Regardless, thanks a lot for the feedback @fauxpark.

**Review by fauxpark — APPROVED** [review]

**Review by drashna — APPROVED** [review]

**Comment by drashna** [comment]
Thanks!

**Pull request merged**